src/wallet/api.rs

```diff
@@ -1255,63 +1260,140 @@ impl Wallet {
+                let tx_weight = 43 + input_weight + target_weight.to_wu();
+                let fee = calculate_fee(tx_weight / 4, feerate as f32).map_err(|e| {
+                    WalletError::General(format!("Group fee calculation failed: {e}"))
+                })?;
+                Ok(fee)
+            };
 
-        for (utxo, spend_info) in &unspents {
+        // Smart Address Grouping
+        let mut address_groups: HashMap<String, Vec<(ListUnspentResultEntry, UTXOSpendInfo)>> =
+            HashMap::new();
+        for (utxo, spend_info) in unspents {
             let address_str = utxo
                 .address
                 .as_ref()
                 .map(|addr| addr.clone().assume_checked().to_string())
                 .unwrap_or_else(|| format!("script_{}", utxo.script_pub_key));
+            address_groups
+                .entry(address_str)
+                .or_default()
+                .push((utxo, spend_info));
+        }
 
-            match address_groups
-                .iter_mut()
-                .find(|(addr, _)| addr == &address_str)
-            {
-                Some(group) => group.1.push((utxo.clone(), spend_info.clone())),
-                None => {
-                    address_groups.push((address_str, vec![(utxo.clone(), spend_info.clone())]))
-                }
+        let (mut grouped_addresses, single_addresses): (Vec<_>, Vec<_>) = address_groups
+            .into_values()
+            .partition(|group| group.len() > 1);
+
+        // Sort for single group optimization
+        grouped_addresses.sort_by_key(|group| {
+            group
+                .iter()
+                .map(|(u, _): &(ListUnspentResultEntry, UTXOSpendInfo)| u.amount.to_sat())
+                .sum::<u64>()
+        });
+
+        // First, try to find single group that covers target+fees
+        if let Some(single_group) = grouped_addresses.iter().find(|group| {
+            let group_total: u64 = group
+                .iter()
+                .map(|(u, _): &(ListUnspentResultEntry, UTXOSpendInfo)| u.amount.to_sat())
+                .sum();
+            let estimated_fee = estimate_group_fee(group).unwrap_or(0);
+            group_total >= target_amount + estimated_fee
+        }) {
+            let group_total: u64 = single_group
+                .iter()
+                .map(|(u, _): &(ListUnspentResultEntry, UTXOSpendInfo)| u.amount.to_sat())
+                .sum();
+            let estimated_fee = estimate_group_fee(single_group).unwrap_or(0);
+            log::info!("Address grouping: Single group solution - {} UTXOs (total: {} sats, target+fee: {} sats)", 
+                single_group.len(), group_total, target_amount + estimated_fee);
+            return Ok(single_group.clone());
+        }
+
+        // Multi-group selection: accumulate groups until target+fees is met
+        let mut pre_selected_utxos = Vec::new();
+        let mut pre_selected_total = 0u64;
+        let mut all_groups = single_addresses;
```

Comment:
There is no need to combine the single and grouped utxos here. Either the loop will return the groups, or coin_selection will happen on the single addresses.